doc/manual/typing.md

```diff
@@ -0,0 +1,629 @@
+of delayed dynamic type errors that static typing was supposed to solve in the
+first place! Remember, the typical problem is the caller passing an value of the
+wrong type that eventually raises an error from within `filter`.
+
+Hopefully, Nickel does have a mechanism to prevent this from happening and to
+provide good error reporting in this situation.  Let us see by ourselves:
+
+```nickel
+lists.filter (fun x => if x % 2 == 0 then x else null) [1,2,3,4,5,6]
+```
+
+Result:
+```
+error: Blame error: contract broken by the caller.
+  ┌─ :1:17
+  │
+1 │ forall a. (a -> Bool) -> List a -> List a
+  │                 ---- expected return type of a function provided by the caller
+  │
+  ┌─ repl-input-45:1:67
+  │
+1 │ lists.filter (fun x => if x % 2 == 0 then x else null) [1,2,3,4,5,6]
+  │                                                                   - evaluated to this expression
+  │
+  = This error may happen in the following situation:
+    1. A function `f` is bound by a contract: e.g. `(Num -> Num) -> Num`.
+    2. `f` takes another function `g` as an argument: e.g. `f = fun g => g 0`.
+    3. `f` is called by with an argument `g` that does not respect the contract: e.g. `f (fun x => false)`.
+  = Either change the contract accordingly, or call `f` with a function that returns a value of the right type.
+  = Note: this is an illustrative example. The actual error may involve deeper nested functions calls.
+
+note:
+    ┌─ <stdlib/lists>:160:14
+    │
+160 │     filter : forall a. (a -> Bool) -> List a -> List a
+    │              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ bound here
+
+[...]
+note:
+  ┌─ repl-input-45:1:1
+  │
+1 │ lists.filter (fun x => if x % 2 == 0 then x else null) [1,2,3,4,5,6]
+  │ -------------------------------------------------------------------- (3) calling <func>
+```
+
+We call `filter` from a dynamically typed location, but still get a spot-on
+error. To precisely avoid dynamically code injecting values of the wrong type
+inside statically typed blocks via function calls, the interpreter protects said
+blocks by a contract. Contracts form a principled runtime verification scheme.
+Please refer to the [dedicated document](./contracts.md) for more details, but
+for now, you can just remember that *any type annotation* (wherever it is) gives
+rise at runtime to a corresponding contract application. In other words, `foo:
+T` and `foo | T` (here `|` is contract application, not the row tail separator)
+behave exactly the same at *runtime*.
+
+Thanks to this guard, you can statically type your library functions and use
+them from dynamically typed code while still enjoying good error messages.
+
+### Using untyped code inside typed code
+
+In the other direction, we face a different issue. How can we use untyped values
+inside typed code? Let us try:
+
+```
+let x = 1 in
+(1 + x : Num)
+```
+
+Although `x` shouldn't be statically typed, this code is accepted. Hmm. What
```

Comment:
I think that this example should be lower down. It's a facility that Nickel offers, but not an essential feature. And it makes the discussion more confused at this point.